Print settings options such evmVersion and optimize in isoltest/soltest failed test logs.

matheusaaguiar opened the pull request:
Resolves #13065.

matheusaaguiar committed:
Added printSelectedOptions and toString as method of CommonOptions and used the former to print settings in failed semanticTests.

cameel commented on test/tools/isoltest.cpp:
```diff
@@ -146,6 +152,16 @@ class TestTool
+{
+	_stream << _linePrefix << "// ====" << endl;
+	_stream << _linePrefix << "// evmVersion: " << m_options.evmVersion().name() << endl;
+	_stream << _linePrefix << "// optimize: " << m_options.optimize << endl;
+	_stream << _linePrefix << "// useABIEncoderV1: " << m_options.useABIEncoderV1 << endl;
```
This will print `0` or `1`, right? `true`/`false` would be better.

cameel commented on test/tools/isoltest.cpp:
```diff
@@ -146,6 +152,16 @@ class TestTool
+	_stream << _linePrefix << "// evmVersion: " << m_options.evmVersion().name() << endl;
+	_stream << _linePrefix << "// optimize: " << m_options.optimize << endl;
+	_stream << _linePrefix << "// useABIEncoderV1: " << m_options.useABIEncoderV1 << endl;
+	_stream << _linePrefix << "// batches: " << m_options.batches << endl;
+	_stream << _linePrefix << "// selectedBatch: " << m_options.selectedBatch << endl;
```
The isoltest output is already pretty verbose since it prints the whole source. I'd try to make this more compact.

For example I'd make this look something like: `batch: 3/125` instead having each part on a separate line.

BTW, what are these set to when not using the `--batches` option? Is this `1`/`1` or zeros or something else?

matheusaaguiar replied:
I guess it would output something like `0/1` if not using --batches. That's what I got when running isoltest locally with `./test/tools/isoltest --enforce-gas-cost --optimize --no-smt`:
```
...
 // ====
    // evmVersion: london
    // optimize: 1
    // useABIEncoderV1: 0
    // batches: 1
    // selectedBatch: 0
```

cameel replied:
I think we should add 1 to batch number in that case. I.e. so that's `1/1` by default and so that it's normally from `1/n` to `n/n` rather than from `0/n` to `n-1/n`.

cameel commented on test/tools/isoltest.cpp:
```diff
@@ -146,6 +152,16 @@ class TestTool
 
+void TestTool::printOptions(ostream& _stream, string const& _linePrefix, bool) const
+{
+	_stream << _linePrefix << "// ====" << endl;
+	_stream << _linePrefix << "// evmVersion: " << m_options.evmVersion().name() << endl;
+	_stream << _linePrefix << "// optimize: " << m_options.optimize << endl;
+	_stream << _linePrefix << "// useABIEncoderV1: " << m_options.useABIEncoderV1 << endl;
+	_stream << _linePrefix << "// batches: " << m_options.batches << endl;
+	_stream << _linePrefix << "// selectedBatch: " << m_options.selectedBatch << endl;
```
This format makes it look like the expectations you can add in the test files but you cannot use it there. I guess we can do it like this too but I was really thinking about something more compact since these values are just extra information that you do not always need. They should not be some prominent.

Maybe it would be better to have them on a single line? I.e. something like:
```
Run settings: EVM=london, optimize, batch=3/125
```
with boolean values like `optimize` and `abiencoderv1` just being omitted when false.

Or, alternatively:
```
Run settings: evmVersion=london, optimize=true, abiencoderv1=false, batch=3/125
```

matheusaaguiar replied:
I would prefer the second form because you leave no room for doubt about the omitted/false values...

cameel replied:
Sure, I'm fine with either.

cameel commented on test/tools/isoltest.cpp:
```diff
@@ -178,6 +194,7 @@ TestTool::Result TestTool::process()
 						AnsiColorized(cout, formatted, {BOLD, CYAN}) << "  Contract:" << endl;
 						m_test->printSource(cout, "    ", formatted);
 						m_test->printSettings(cout, "    ", formatted);
+						printOptions(cout, "    ", formatted);
```
Can you make sure that `isoltest` won't add this to the rewritten test file when the test fails and it offers to update the expectations for you and you press `u` to accept?

matheusaaguiar replied:
Sorry, I didn't quite understand what you meant here ... Do you mean preventing someone in the future doing this? Or do you mean making sure that it can't be reached from inside method `updateCase` ?

cameel replied:
Yes, I mean `updateTestCase()`.

Have you seen what happens when you run `isoltest` and the expectations do not match? `soltest` just fails, but `isoltest` stops and asks you what to do. One of the options is to rewrite the test case with new expectations. I just wanted you to make sure this part is only printed on the screen but not included in the rewritten file (because it would fail then).

matheusaaguiar replied:
Ah, ok!
Yes, I am sure it won't update the files. I have tested locally, including updating a failed test (both with --accept-updates and manually pressing `u` when prompted).

cameel replied:
ok, sounds good then.

cameel commented on test/tools/isoltest.cpp:
```diff
@@ -146,6 +152,16 @@ class TestTool
 
 bool TestTool::m_exitRequested = false;
 
+void TestTool::printOptions(ostream& _stream, string const& _linePrefix, bool) const
```
Why include `_formatted` if you're not using it? It does not look like you're overriding this function.
```suggestion
void TestTool::printOptions(ostream& _stream, string const& _linePrefix) const
```

matheusaaguiar replied:
ops, forgot to delete that ...

cameel commented on test/tools/isoltest.cpp:
```diff
@@ -146,6 +151,15 @@ class TestTool
 
+void TestTool::printOptions(ostream& _stream, string const& _linePrefix) const
+{
+	_stream << _linePrefix << "// Run settings:" <<
+				" EVM=" << m_options.evmVersion().name() <<
+				", optimize=" << (m_options.optimize ? "true" : "false") <<
+				", useABIEncoderV1=" << (m_options.useABIEncoderV1 ? "true" : "false") <<
+				", batch=" << (m_options.selectedBatch+1) << "/"  << m_options.batches << endl;
```
```suggestion
	_stream << _linePrefix << "// Run settings:" <<
		" EVM=" << m_options.evmVersion().name() <<
		", optimize=" << (m_options.optimize ? "true" : "false") <<
		", useABIEncoderV1=" << (m_options.useABIEncoderV1 ? "true" : "false") <<
		", batch=" << (m_options.selectedBatch+1) << "/"  << m_options.batches << endl;
```

cameel replied:
```suggestion
	_stream << _linePrefix << "// Run settings:" <<
				" EVM=" << m_options.evmVersion().name() <<
				", optimize=" << (m_options.optimize ? "true" : "false") <<
				", useABIEncoderV1=" << (m_options.useABIEncoderV1 ? "true" : "false") <<
				", batch=" << (m_options.selectedBatch + 1) << "/" << m_options.batches << endl;
```

cameel commented on test/tools/isoltest.cpp:
```diff
@@ -115,6 +115,11 @@ class TestTool
 
 	Result process();
 
+	void printOptions(
+		ostream& _stream,
+		string const& _linePrefix = ""
+	) const;
```
This is short enough to be on a single line now.
```suggestion
	void printOptions(ostream& _stream, string const& _linePrefix = "") const;
```

cameel reviewed (commented):
Just minor style fixes left.

If the thing with expectations I mentioned earlier works fine, it should be good to go.

As the final test though, could you add some errors to expectations so that we can see how the output looks like now in CI? That was the primary motivation for this after all. Would be good to see how it looks when multiple processes are printing output simultaneously. You can just add a commit with an error, let it do a complete run and then remove that commit and just post a link to the failed run.

EDIT: Also, I guess it would be best to rebase the PR on #13072 first so that the output actually gets captured (unless we just merge it before then).

cameel commented:
I just merged #13072 so when you're done you can simply rebase on `develop`.

matheusaaguiar commented:
Ok.

Failed test run: [5b537588-6425-40b3-92e5-05a5414aff8a](https://app.circleci.com/pipelines/github/ethereum/solidity/24435/workflows/5b537588-6425-40b3-92e5-05a5414aff8a).

Rebasing and removing errors now.

cameel commented:
> Failed test run: [5b537588-6425-40b3-92e5-05a5414aff8a](https://app.circleci.com/pipelines/github/ethereum/solidity/24435/workflows/5b537588-6425-40b3-92e5-05a5414aff8a).

Looks like something's wrong. I can't see the extra output in any of the failed runs.

matheusaaguiar commented:
I am investigating.

matheusaaguiar commented:
Changed the way the options information is added to test failure messages. Added it to the end of the `outputMessage` returned by the test and colored it yellow.I think it will show in CI now.

matheusaaguiar commented:
I figured out what happened. I only looked at isoltest and forgot about soltest. So the changes I made were only affecting the former. So here's what I did: I implemented method `toString()` and helper function `printOptions` in `CommonOptions`.  I think this is a good way to be able to access the options values in both testers, isoltest and soltest, or any other test that uses `CommonOptions`.

matheusaaguiar commented:
Now the options are showing in CI failed tests also: [076cf593-a2f5-4e95-9609-7aa537b52855](https://app.circleci.com/pipelines/github/ethereum/solidity/24466/workflows/076cf593-a2f5-4e95-9609-7aa537b52855)

matheusaaguiar commented on test/Common.cpp:
```diff
@@ -207,6 +207,74 @@ bool CommonOptions::parse(int argc, char const* const* argv)
+	{
+		switch (options) {
+			case "evmVersion":
+				options << "evmVersion=" << evmVersion().name() << " ";
```
I think adding a functionality like this could be helpful in the future if we want to print other options... But I don't know if this is a probable scenario or if it is worth the effort... some inputs about this would be appreciated.

matheusaaguiar commented:
Rebased on develop.
I am thinking that maybe in the future we could want to print other options. Is it worth to implement a way of specifying what option values are desired in the `printOptions` functions ? Maybe use std::initializer_list or a STL container like std::vector?
The thing that is bugging me is that the way it is now, it is just boilerplate code...

cameel commented on test/Common.cpp:
```diff
@@ -207,6 +207,38 @@ bool CommonOptions::parse(int argc, char const* const* argv)
+		{"showMetadata", "showMetadata=" + boolToString(showMetadata)}
+	};
+
+	for(auto optionName : _filter)
```
Our style is to always but a space betwene `if` and `(` or `for` and `(`.
```suggestion
	for (auto optionName: _filter)
```

cameel commented on test/Common.cpp:
```diff
@@ -207,6 +207,41 @@ bool CommonOptions::parse(int argc, char const* const* argv)
+		return "";
+
+	//Using std::map to avoid if-else/switch-case block
+	std::map<std::string, std::string> option_value_map = {
```
```suggestion
	std::map<std::string, std::string> optionValueMap = {
```

cameel commented on test/Common.cpp:
```diff
@@ -207,6 +207,41 @@ bool CommonOptions::parse(int argc, char const* const* argv)
+		{"evmVersion", "evmVersion=" + evmVersion().name()},
+		{"optimize", "optimize=" + boolToString(optimize)},
+		{"useABIEncoderV1", "useABIEncoderV1=" + boolToString(useABIEncoderV1)},
+		{"batch", "batch=" + std::to_string(selectedBatch+1) + "/" + std::to_string(batches)},
```
```suggestion
		{"batch", "batch=" + std::to_string(selectedBatch + 1) + "/" + std::to_string(batches)},
```

cameel commented on test/Common.cpp:
```diff
@@ -207,6 +207,41 @@ bool CommonOptions::parse(int argc, char const* const* argv)
+
+	//erase extra ', '
+	if(!options.empty())
+		options.erase(options.size()-2);
```
You should use ranges v3 and our `joinHumanReadable()` here.

cameel commented on test/Common.cpp:
```diff
@@ -207,6 +207,41 @@ bool CommonOptions::parse(int argc, char const* const* argv)
 	return true;
 }
 
+std::string CommonOptions::toString(std::vector<std::string> const& _filter) const
```
Let's just add `using namespace std;` and drop the `std::` prefixes. We do this everywhere.

cameel commented on test/Common.cpp:
```diff
@@ -207,6 +207,41 @@ bool CommonOptions::parse(int argc, char const* const* argv)
+	if(!m_singleton || _filter.empty())
+		return "";
+
+	//Using std::map to avoid if-else/switch-case block
```
```suggestion
	// Using std::map to avoid if-else/switch-case block
```

cameel commented on test/Common.cpp:
```diff
@@ -277,4 +312,14 @@ bool loadVMs(CommonOptions const& _options)
+	_stream << _linePrefix << "Run Settings: " << _options.toString(_filter) << std::endl;
+}
+
+void printOptions(std::ostream& _stream, std::string const& _linePrefix, CommonOptions const& _options)
```
It would make more sense to call this something like `printImportantOptions()`. Without the argument I'd actually expect it to print everyting, not some random subset.
```suggestion
void printImportantOptions(std::ostream& _stream, std::string const& _linePrefix, CommonOptions const& _options)
```

matheusaaguiar replied:
I have changed it to `printSelectedOptions`.

cameel commented on test/Common.h:
```diff
@@ -102,6 +107,13 @@ bool isValidSemanticTestPath(boost::filesystem::path const& _testPath);
+void printOptions(std::ostream& _stream, std::string const& _linePrefix, CommonOptions const& _options, std::vector<std::string> const& _filter);
+void printOptions(std::ostream& _stream, std::string const& _linePrefix, CommonOptions const& _options);
+
+inline std::string const boolToString(bool b) { return b ? "true" : "false"; }
```
I'd either make this a lambda for local use in `toString()` or put it in [`CommonData.h`](https://github.com/ethereum/solidity/blob/develop/libsolutil/CommonData.h) if you really think it makes sense to make it generic.

matheusaaguiar replied:
I guess a lambda is what I was looking for since the beginning, so I'll go with it.

cameel commented on test/Common.h:
```diff
@@ -102,6 +107,13 @@ bool isValidSemanticTestPath(boost::filesystem::path const& _testPath);
+void printOptions(std::ostream& _stream, std::string const& _linePrefix, CommonOptions const& _options, std::vector<std::string> const& _filter);
+void printOptions(std::ostream& _stream, std::string const& _linePrefix, CommonOptions const& _options);
+
+inline std::string const boolToString(bool b) { return b ? "true" : "false"; }
```
```suggestion
inline std::string const boolToString(bool value) { return value ? "true" : "false"; }
```

cameel commented on test/libsolidity/semanticTests/strings/empty_storage_string.sol:
```diff
@@ -84,9 +84,9 @@ contract C {
 // i() -> 0x40, 0x80, 0x1a, 38178759162904981154304545770567765692299154484752076569098748838215919075328, 0
-// j(string): 0x20, 0, "" -> 0x20, 0
+// j(string): 0x80, 0, "" -> FAILURE
 // k() -> 0x20, 0
-// l(string): 0x20, 0, "" -> 0x20, 0x40, 0x20, 0
+// l(string): 0x40, 0, "" -> 0x20, 0x40, 0x20, 0
```
This looks unrelated to this PR. Why is it here?

matheusaaguiar replied:
I guess it has to do with local tests and I must have overlooked when staging and committing. I am going to restore to the correct version.

cameel commented on test/Common.h:
```diff
@@ -102,6 +107,13 @@ bool isValidSemanticTestPath(boost::filesystem::path const& _testPath);
 
+///Helper function to print the value of settings used (CommonOptions struct)
+///@p _filter should be the sequence of option names to be printed.
+void printOptions(std::ostream& _stream, std::string const& _linePrefix, CommonOptions const& _options, std::vector<std::string> const& _filter);
```
Maybe `_selectedOptions` would be more self-explanatory than just `_filter`?
```suggestion
void printOptions(std::ostream& _stream, std::string const& _linePrefix, CommonOptions const& _options, std::vector<std::string> const& _selectedOptions);
```

cameel commented on test/Common.h:
```diff
@@ -102,6 +107,13 @@ bool isValidSemanticTestPath(boost::filesystem::path const& _testPath);
 
 bool loadVMs(CommonOptions const& _options);
 
+///Helper function to print the value of settings used (CommonOptions struct)
+///@p _filter should be the sequence of option names to be printed.
+void printOptions(std::ostream& _stream, std::string const& _linePrefix, CommonOptions const& _options, std::vector<std::string> const& _filter);
+void printOptions(std::ostream& _stream, std::string const& _linePrefix, CommonOptions const& _options);
```
These take `CommonOptions` as an argument. It seems to me like they should just be its methods rather than standalone functions.

cameel commented:
> I am thinking that maybe in the future we could want to print other options. 

Maybe, but I'd avoid overgeneralizing it. Adding a new option would not have been a major effort even with the code you originally had here. But since you have already written it, we can just keep the general version now :)

> Is it worth to implement a way of specifying what option values are desired in the printOptions functions ? Maybe use std::initializer_list or a STL container like std::vector?
> The thing that is bugging me is that now it is just hardcoded to some options...

A vector seems fine. If we wanted it to be really clean, I'd use an enum instead of strings for specifying options. But like I said, not much point extending it too much. What you have now is fine.

I'd keep only the version of `printOptions()` that takes the list of options and I'd hard-code the options we want at the point of call. After all the ones we want are specific to our use case (i.e. debugging a semantic test). If you want to avoid had-coding same options in two places maybe just create a constant for the default vector.

cameel commented on test/boostTest.cpp:
```diff
@@ -108,6 +107,7 @@ void runTestCase(TestCase::Config const& _config, TestCase::TestCaseCreator cons
 				case TestCase::TestResult::Success:
 					break;
 				case TestCase::TestResult::Failure:
+					solidity::test::printOptions(errorStream, "", solidity::test::CommonOptions::get());
```
Are you sure that printing here will make these options show up only for semantic tests? That's the only situation in which the test result may be affected by them so it does not make much sense to print them otherwise.

matheusaaguiar replied:
You are right, I lost sight of that important detail and was thinking about all tests. I have already moved it to only print in failed semantic tests.

matheusaaguiar commented:
I have implemented all your suggestions @cameel.

cameel commented on test/Common.cpp:
```diff
@@ -207,6 +211,45 @@ bool CommonOptions::parse(int argc, char const* const* argv)
+	if (!m_singleton || _selectedOptions.empty())
+		return "";
+
+	auto boolToString = [](bool value) -> string
+		{
+			return value ? "true" : "false";
+		};
```
```suggestion
	auto boolToString = [](bool value) -> string
	{
		return value ? "true" : "false";
	};
```

But it's short enough that I'd just put it on a single line.

cameel commented on test/Common.cpp:
```diff
@@ -207,6 +211,45 @@ bool CommonOptions::parse(int argc, char const* const* argv)
+	};
+
+	using namespace ranges;
+	vector<string> options = _selectedOptions |
+		views::transform([&optionValueMap](auto& optionName) { return optionValueMap.find(optionName); }) |
+		views::filter([&optionValueMap](auto it) { return it != optionValueMap.end(); }) |
+		views::transform([](auto it) { return it->second; }) |
+		to<vector>();
```
1) I just noticed that you just ignore options that are not present in `optionValueMap`. I think we should require them to be present instead. Option names are not user input so we do not need to handle them gracefully and in practice, when you make a typo in the name it's better to get an error than for it to be silently skipped.
2) Since the `option=` part is always identical to the key, why not remove it from the map and add it here? Less duplication and the map would be more readable.

```suggestion
	soltestAssert(views::set_difference(_selectedOptions, optionValueMap | views::keys).empty());

	vector<string> optionsWithValues = _selectedOptions |
		views::transform([&optionValueMap](string const& _option) { return _option + "=" + optionValueMap.at(_option); }) |
		to<vector>();
```

matheusaaguiar replied:
1. Yes, it really makes sense to do so. I did not think about erroneous values at the time I came up with that code.
2. I should have done it like that since the beginning.
Thanks for the excellent review and suggestions!

matheusaaguiar replied:
I could not use `set_differences` because `_selectedOptions` should be sorted prior to that, but then it is a `const` reference. So I used `all_of` instead...

cameel commented on test/Common.cpp:
```diff
@@ -207,6 +211,45 @@ bool CommonOptions::parse(int argc, char const* const* argv)
+	return solidity::util::joinHumanReadable(options);
+}
+
+void CommonOptions::printSelectedOptions(std::ostream& _stream, std::string const& _linePrefix, std::vector<std::string> const& _selectedOptions) const
+{
+	_stream << _linePrefix << "Run Settings: " << toString(_selectedOptions) << std::endl;
+}
```
```suggestion
void CommonOptions::printSelectedOptions(ostream& _stream, string const& _linePrefix, vector<string> const& _selectedOptions) const
{
	_stream << _linePrefix << "Run Settings: " << toString(_selectedOptions) << endl;
}
```

cameel commented on test/Common.h:
```diff
@@ -81,6 +81,12 @@ struct CommonOptions
 	// Throws a ConfigException on error
 	virtual void validate() const;
 
+	// @returns string with a key=value list of the options separated by comma
+	// Ex.: "evmVersion=london, optimitze=true, useABIEncoderV1=false"
+	virtual std::string toString(std::vector<std::string> const& _selectedOptions) const;
+	///Helper to print the value of settings used
```
```suggestion
	/// @returns string with a key=value list of the options separated by comma
	/// Ex.: "evmVersion=london, optimize=true, useABIEncoderV1=false"
	virtual std::string toString(std::vector<std::string> const& _selectedOptions) const;
	/// Helper to print the value of settings used
```

matheusaaguiar commented on test/Common.cpp:
```diff
@@ -207,6 +212,41 @@ bool CommonOptions::parse(int argc, char const* const* argv)
+		{"showMetadata", boolToString(showMetadata)}
+	};
+
+	soltestAssert(ranges::all_of(_selectedOptions, [&optionValueMap](string const& _option) { return optionValueMap.find(_option) != optionValueMap.end(); }));
```
I could not use `set_differences` because `_selectedOptions` should be sorted prior to that, since the `optionValueMap` keys are sorted. But then, it is a `const` reference,  so I used `all_of` instead...

cameel replied:
Ah, you probably need `_selectedOptions | to<set>()` for it to work. I usually find the set versions more concise but either is fine really.

You could just use `count()` to shorten it since you don't use the iterator returned by `find()` for anything:
```suggestion
	soltestAssert(ranges::all_of(_selectedOptions, [&optionValueMap](string const& _option) { return optionValueMap.count(_option) > 0; }));
```

matheusaaguiar commented:
Rebased on develop.

cameel commented on test/Common.cpp:
```diff
@@ -207,6 +212,41 @@ bool CommonOptions::parse(int argc, char const* const* argv)
+string CommonOptions::toString(vector<string> const& _selectedOptions) const
+{
+	if (!m_singleton || _selectedOptions.empty())
+		return "";
```
This function operates on the current instance of options and does not access `m_singleton`. Is this condition actually needed?

The ` _selectedOptions.empty()` part is fine but it seems that function would still work fine without it so I'd remove it. It's such a useless case that it's not really worth optimizing for.

matheusaaguiar replied:
Yes, I should have changed that once I made it become a member of the struct.

cameel reviewed (commented):
Two more tweaks and we're done here.
